Optimization of association analysis.  Command line interface.  Bug fixes.  Code cleanup.

andychu opened the pull request:
- Added bin/decode-assoc command line tool for automation.  Define and read a new schema so we know whether variables are string or boolean, and what their encoding parameters are.
- bin/test.sh: Add tests for bin/decode-assoc, for (string x boolean) problem
  - Add Boolean RAPPOR encoding  (no hashing step) to the Python client with Encoder.encode_bits
  - Add association testdata generation to rappor_sim.py

- Optimizations
  - In association.R, remove from GetCondProb() any computation that doesn't depend on the report.  (e.g. 100x-1000x speedup for the joint conditional stage)
  - Use mclapply in R for the steps that did lapply() over all reports (N).  The number of cores is controlled by decode-assoc --num-cores.
  - Provide an alternative implementation of EM in C++ in analysis/cpp/fast_em.cc (analysis/R/fast_em.R is a wrapper that does serialization and a system() call)  ~100x speedup.

- Fixed bugs in the k=1 case, due to R matrix/vector confusion (adapted from Ananth's changes)
- Allow different variables in association analysis to have different parameters (adapted from Ananth's changes, but tests still pass)
- Add hacky support for boolean vars (adapted from Ananth's changes)

- Code cleanup in association.R.  Rename variables to be more clear.
- Minor refactoring of decode_dist.R to resemble decode_assoc.R

tkaitchuck commented on analysis/R/association.R:
```diff
@@ -21,94 +31,100 @@ GetOtherProbs <- function(counts, map, marginal, params) {
-  #       "other" category)
   #   params: System parameters
-  #   map: list of matrices encoding locations of hashes for each string
+  #   bit_indices: list of integer vectors of length h, specifying which bits
```
This line should indicate that these are canidates

tkaitchuck commented on analysis/R/association.R:
```diff
@@ -296,34 +318,59 @@ ComputeDistributionEM <- function(reports, report_cohorts,
-  #   params: A list of parameters for the problem
+  #   params: RAPPOR encoding parameters.  If set, all variables are assumed to
+  #       be encoded with these parameters.
+  #   params_list: A list of RAPPOR encoding parameters (each of which are
```
Why do we have both params and params_list?

tkaitchuck commented on analysis/R/association.R:
```diff
@@ -296,34 +318,59 @@ ComputeDistributionEM <- function(reports, report_cohorts,
   if (num_variables > 4) {
-    cat("This is too many variables to compare, exiting now.")
-    return -1
+    stop("Not comparing more than 4 variables")
```
The error message should print num_variables

tkaitchuck commented on analysis/R/association.R:
```diff
@@ -332,38 +379,67 @@ ComputeDistributionEM <- function(reports, report_cohorts,
+    pstar <- (1 - f / 2) * p + (f / 2) * q
+    qstar <- (1 - f / 2) * q + (f / 2) * p
+
+    bit_indices_by_cohort <- lapply(1:var_params$m, function(cohort) {
```
Did you test using mclapply here?

tkaitchuck commented on analysis/R/association.R:
```diff
@@ -147,11 +163,15 @@ UpdatePij <- function(pij, cond_prob) {
   # Returns:
   #   Updated pijs from em algorithm (maximization)
 
+  # NOTE: Not using mclapply here because we have a faster C++ implementation.
```
This is already being called inside an mclapply. So the better reason why it should not be is because it would be redundant.

tkaitchuck commented on analysis/R/fast_em.R:
```diff
@@ -0,0 +1,104 @@
+
+ConstructFastEM <- function(em_executable, tmp_dir) {
+  # TODO:
+  # - check that number of dimensions is 2
```
Add this check

tkaitchuck commented on analysis/cpp/fast_em.cc:
```diff
@@ -0,0 +1,294 @@
+  }
+
+  // Now read dynamic data
+  uint32_t vec_length = num_entries * entry_size;
```
You should really put bounds on these things so there is no possibility of wrapping here.

tkaitchuck commented on analysis/cpp/run.sh:
```diff
@@ -7,14 +7,35 @@ set -o nounset
-  g++ -std=c++0x -o find_cliques find_cliques.cc 
+# Call gcc with the flags we like.
+
+# -O3: Optimize yet more. -O3 turns on all optimizations specified by -O2 and
```
We don't need this comment section

tkaitchuck commented on bin/test.sh:
```diff
@@ -29,4 +40,125 @@ decode-dist() {
+  decode-assoc --em-executable "$EM_EXECUTABLE"
+}
+
+# TODO: Compare these results somehow?  Or just eyeball them.
```
We should compare these two. They really should be identical modulo precision in the math operations.

andychu commented:
OK thanks for the review.  I addressed all but the last one.  I will do something to compare the results.  I think the problem is that Decode() is fairly noisy on a small number of reports.

DONE: Updated docstring for bit_indices

Having both params and params_list is mainly because the unit tests already use a single set of params.

I removed the num_variables > 4 check, because it was arbitrary.  The unit tests test 3 variables, so 4 should work.

I only used mclapply() in cases where we're iterating over the number of reports, but not say the number of cohorts.  mclapply() forks processes so it has some significant memory overhead, as well as startup overhead.

UpdatePij: This is being called inside for (i in 1:max_em_iters), not mclapply?

Added the check that dimensions == 2 in ConstructFastEM.

Wrapping: I changed this variable type to size_t.  num_entries and entry_size are uint32_t, but their product could be more than 4 GB, which should work fine on a 64 bit machine.  I don't want to hard-code any assumptions about the machine memory size.

analysis/cpp/run.sh -- removed extra comments

andychu commented:
OK, in association_test.R, I added TestTwoImplementations() which runs both algorithms.  The L1 difference between the matrices is 9.7e-17 (not 0 curiously).

I also renamed the ugly $rmap and $map stuff, and did a few other cleanups.

PTAL.

ananthr commented on bin/decode_assoc.R:
```diff
@@ -0,0 +1,350 @@
+
+  fit_df <- as.data.frame(fit)
+  print(fit_df)
+
```
Hey Andy, for now, would you like to use the trick and rename Other to FALSE for boolean so that it is consistent with internal reports?

I tested out the code with the foll. to produce internal reports that are consistent with what I did manually.

rownames(fit_df)[which(rownames(fit_df) == "Other")] = "FALSE"

ananthr commented on bin/decode_assoc.R:
```diff
@@ -0,0 +1,350 @@
+  print(fit_df)
+
+  results_csv_path <- file.path(opts$output_dir, 'assoc-results.csv')
+  write.csv(fit_df, file = results_csv_path)
```
And can we write melt(t(fit_df)) 

melt() requires reshape2 library. The advantage of this is that the data frame is flattened and we can extend our results to 3 or higher dimensional associations naturally. Tables only work for 2 dimensions.

write.csv(melt(t(fit_df)), file = results_csv_path, quote = FALSE, row.names = FALSE) would be consistent with internal reports.

andychu commented:
OK good point.  I changed this and the CSV now looks like this:

(I ended up using something a little different than "melt", since that was doing weird things)

"flag..HTTPS","domain","proportion"
"TRUE","bing.com",0.263930065623489
"FALSE","bing.com",0.147970233934715
"TRUE","yahoo.com",0.182032391377609
"FALSE","yahoo.com",0.114596102564608
"TRUE","google.com",0.190706050320775
"FALSE","google.com",0.095164303098314
"TRUE","Other",0.00340437927526885
"FALSE","Other",0.00219647380522421

andychu commented:
Any more comments on this?

ananthr commented:
[LGTM]

Thanks Andy! Curious to know what problems you have with melt, but anything that flattens the assoc results before printing is sufficient for now.
